src/box/iproto.cc

```diff
@@ -2701,22 +2722,211 @@ tx_process_replication(struct cmsg *m)
+	assert(k != mh_end(tx_req_handlers));
+	mh_i32ptr_del(tx_req_handlers, k, NULL);
+}
+
+/**
+ * Replaces an event in `handlers' by the new `event'. If `is_by_id', the
+ * handler is set by request type id, otherwise it is set by request type name.
+ */
+static void
+iproto_req_handlers_set_event(struct iproto_req_handlers *handlers,
+			      struct event *event, bool is_by_id)
+{
+	assert(handlers != NULL);
+	assert(event != NULL);
+
+	event_ref(event);
+
+	if (is_by_id) {
+		if (handlers->event_by_id != NULL)
+			event_unref(handlers->event_by_id);
+		handlers->event_by_id = event;
+	} else {
+		if (handlers->event_by_name != NULL)
+			event_unref(handlers->event_by_name);
+		handlers->event_by_name = event;
+	}
+}
+
+/**
+ * Deletes an event, which is set in `handlers' by request type id (if
+ * `is_by_id'), or by request type name.
+ */
+static void
+iproto_req_handlers_del_event(struct iproto_req_handlers *handlers,
+			      bool is_by_id)
+{
+	assert(handlers != NULL);
+
+	if (is_by_id) {
+		event_unref(handlers->event_by_id);
+		handlers->event_by_id = NULL;
+	} else {
+		event_unref(handlers->event_by_name);
+		handlers->event_by_name = NULL;
+	}
+}
+
+/**
+ * Returns `true' if there is at least one handler in `handlers'.
+ */
+static bool
+iproto_req_handler_is_set(struct iproto_req_handlers *handlers)
+{
+	if (handlers == NULL)
+		return false;
+
+	return handlers->event_by_id != NULL ||
+	       handlers->event_by_name != NULL ||
+	       handlers->c.cb != NULL;
+}
+
+/**
+ * Returns `enum iproto_type' if `name' is a valid IPROTO type name or equals
+ * "unknown". Otherwise, iproto_type_MAX is returned. The name is expected to
+ * be in lowercase.
+ */
+static enum iproto_type
+get_iproto_type_by_name(const char *name)
+{
+	for (uint32_t i = 0; i < iproto_type_MAX; i++) {
+		const char *type_name = iproto_type_name_lower(i);
+		if (type_name != NULL && strcmp(type_name, name) == 0)
+			return (enum iproto_type)i;
+	}
+	if (strcmp(name, "unknown") == 0)
+		return IPROTO_UNKNOWN;
+	return iproto_type_MAX;
+}
+
+/**
+ * Runs triggers registered for the `event'.
+ * The given header and body the IPROTO packet are passed as trigger args.
+ * Returns IPROTO_HANDLER_OK if some trigger successfully handled the request,
+ * IPROTO_HANDLER_FALLBACK if no triggers handled the request, or
+ * IPROTO_HANDLER_ERROR on failure.
+ */
+static enum iproto_handler_status
+tx_run_override_triggers(struct event *event, const char *header,
+			 const char *header_end, const char *body,
+			 const char *body_end)
+{
+	enum iproto_handler_status rc = IPROTO_HANDLER_FALLBACK;
+	const char *name = NULL;
+	struct func_adapter *trigger = NULL;
+	struct func_adapter_ctx ctx;
+	struct event_trigger_iterator it;
+	event_trigger_iterator_create(&it, event);
+
+	while (event_trigger_iterator_next(&it, &trigger, &name)) {
+		struct mp_ctx mp_ctx_header, mp_ctx_body;
+		mp_ctx_create_default(&mp_ctx_header, iproto_key_translation);
```

Comment:
I'm not sure about this part. Is it necessary to call `mp_ctx_create_default()` on every iteration?

Reply:
Also, can we use one `mp_ctx` for both packets?

Reply:
Yes.

Reply:
Actually it doesn't work, because `func_adapter_push_msgpack_with_ctx` calls `mp_ctx_move`.